WooCommerce/Classes/ViewRelated/Dashboard/Settings/Help/ApplicationLogViewModel.swift

```diff
@@ -0,0 +1,130 @@
+import Foundation
+import Combine
+import UIKit
+
+struct ApplicationLogLine: Identifiable {
+    let id = UUID()
+
+    let date: Date?
+    let text: String
+
+    private let dateFormatter: DateFormatter = {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "YYYY-MM-dd HH:mm:ss:SSS"
+        return dateFormatter
+    }()
+
+    private let datePrefixPattern = try! NSRegularExpression(pattern: "^(\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2}\\:\\d{3})  (.*)")
+
+
+    init(text: String) {
+        guard let result = datePrefixPattern.firstMatch(in: text, options: [], range: NSMakeRange(0, text.utf16.count)),
+              let dateRange = Range(result.range(at: 1), in: text),
+              let logRange = Range(result.range(at: 2), in: text),
+              let date = dateFormatter.date(from: String(text[dateRange])) else {
+            self.date = nil
+            self.text = text
+            return
+        }
+
+        self.date = date
+        self.text = String(text[logRange])
+    }
+}
+
+final class ApplicationLogViewModel: ObservableObject {
+    private let logText: String
+    let lines: [ApplicationLogLine]
+
+    let lastLineID: UUID
+
+    var present: ((UIViewController) -> Void)? = nil
+
+    @Published var lastCellIsVisible = false
+
+    @Published var buttonVisible = true
+
+    private var cancellableSet: Set<AnyCancellable> = []
+
+    init(logText: String) {
+        self.logText = logText
+        lines = logText
+            .split(separator: "\n")
+            .map(String.init)
+            .map(ApplicationLogLine.init(text:))
+
+        // The `ScrollViewProxy.scrollTo` method doesn't seem to handle an optional line well
+        // Instead of `nil`, we can represent a missing value with a random UUID
+        lastLineID = lines.last?.id ?? UUID()
+
+        guard lines.isNotEmpty else {
+            // If the file is empty, there will be no last cell,
+            // and so the last cell won't be visible, but it still doesn't make
+            // sense to show a button to scroll down
+            buttonVisible = false
+            return
+        }
+
+        $lastCellIsVisible
+            .map(!)
+            .debounce(for: .milliseconds(100), scheduler: DispatchQueue.main)
+            .assign(to: \.buttonVisible, on: self)
+            .store(in: &cancellableSet)
+    }
+
+    func isLastLine(_ line: ApplicationLogLine) -> Bool {
+        line.id == lastLineID
+    }
+
+    func showShareActivity() {
+        let activityVC = UIActivityViewController(activityItems: [logText], applicationActivities: nil)
+        activityVC.modalPresentationStyle = .popover
+        activityVC.excludedActivityTypes = assembleExcludedSupportTypes()
+        present?(activityVC)
```

Comment:
⚠️  I'm not sure if these view models should be aware of UIKit types.
Do you think having something like an `enum` representing the navigation cases for then sending it in the closure or in a publisher is an overkill?

Reply:
I was hesitant to invest more time in refactoring this, but finally I decided to go all in and implemented a SwiftUI share sheet in b665286e3, which makes for a much nicer API